markup.go

```diff
@@ -154,26 +150,49 @@ func (m ClassMap) Apply(h *HTML) {
 	Property("className", strings.Join(classes, " ")).Apply(h)
 }
 
-// List represents a list of Markup, Component, or HTML which is individually
+// MarkupList represents a list of Applyer which is individually
 // applied to an HTML element or text node.
-type List []MarkupOrComponentOrHTML
+type MarkupList []Applyer
 
-// Apply implements the Markup interface.
-func (l List) Apply(h *HTML) {
-	for _, m := range l {
-		apply(m, h)
+// Apply implements the Applyer interface.
+func (m MarkupList) Apply(h *HTML) {
+	for _, a := range m {
+		if a == nil {
+			continue
+		}
+		a.Apply(h)
 	}
 }
 
+// Markup wraps a list of Applyer which is individually
+// applied to an HTML element or text node.
+func Markup(m ...Applyer) MarkupList {
```

Comment:
I think the return type here should just be `Applyer`.

I think the description should be simplified to just:

```
// Markup returns an applyer that applies all of the given markup.
```

Reply:
I specifically _don't_ want `Applyer` to be a valid type at the top level, that would mess up our ability to work with `nil` correctly.

Reply:
There is still an issue that I forgot to mention - swapping `nil` for a `MarkupList` between renders will cause us problems - I actually considered not exporting `MarkupList` to avoid this, but having unexported return types can be annoying.

Reply:
Having an unexported type be returned sounds a bit hacky.

> I specifically don't want Applyer to be a valid type at the top level, that would mess up our ability to work with nil correctly.

I'd really like to understand this more, could you point me to relevant code where this being an `Applyer` would cause trouble? Or provide a small example demo'ing the problem?

Reply:
`nil` now occupies a child slot, so anything that returned `nil` in place of `Applyer` would break component persistence.

If `Applyer` is valid at the top level (in which case, there's no point having a `Markup` function, and `Applyer` could just stay named `Markup`):
```go
func (c *MyComponent) OptionalMarkup() vecty.Applyer {
        if c.condition {
                return vecty.Property("key", "value")
        }

        return nil
}

func (c *MyComponent) Render() *vecty.HTML {
        return elem.Div(
                c.OptionalMarkup(), // breaks persistence
                &OtherComponent{},
        )
}
```

We currently have a similar problem with `MarkupList`:
```go
func (c *MyComponent) OptionalMarkup() vecty.MarkupList {
        if c.condition {
                return vecty.Markup(vecty.Property("key", "value"))
        }

        return nil
}

func (c *MyComponent) Render() *vecty.HTML {
        return elem.Div(
                c.OptionalMarkup(), // breaks persistence
                &OtherComponent{},
        )
}
```

However, that becomes impossible if `MarkupList` is unexported, since it cannot be used as a return value, so we will never be able to receive a concrete `nil` in it's place, and it forces the use of `vecty.Markup()` for every case.

Reply:
Ahhh I see, understood. I had forgotten that `MarkupList` must effectively sit side-by-side with `Component` and `*HTML`, and that nil `*HTML` occupies a child slot.

This would not be the case of we forced users to supply Markup always:

```
func Div(markup vecty.Markup, children ...vecty.ComponentOrHTML) *vecty.HTML {
    ...
}
```

But understandably that would be a worse API for situations like `elem.Paragraph(vecty.Text("hello world!"))`.

I understand why you proposed an unexported `MarkupList` above, now, and agree that sounds best.